### ISSUE TITLE
Remove `final` keyword from Slim classes

### ISSUE THREAD

**Issue by stevenwadejr** [slimphp/Slim#1491]
Since Slim is a minimalist framework and comes with the bare minimum, it leaves endless possibilities to expand and add functionality to the core. However, since some classes within the core are declared as `final` (`\Slim\CallableResolver`, `\Slim\Container`), it's more difficult and sometimes impossible to override or extend the core to add or alter functionality.

I propose removing the `final` keyword from core Slim classes to allow more flexibility.

**Comment by JoeBengalen** [comment]
Those are the only classes declared final. Both can very easily be replaced.

The `callableResolver` implements the `CallableResolverInterface` which has only a single method. No need to extend that. It can be replaced by registering your own `callableResolver` service (which must implement the same interface) in the container.

The container implements the `Interop\Container\ContainerInterface` and can be replaced by any other container by passing it into the `Slim\App()` constructor. There is nothing in the container why you would need to extend it. Every defined service in there can be replaced or skipped by defining your own version of the service.

**Comment by stevenwadejr** [comment]
Ok, the `CallableResolver` isn't a great example, but I think the container should not be `final`. One could easily implement a new instance of the `ContainerInterface`, but then you'd have to re-implement all of the configuration setup by `Slim\Container`. Without `final`, one could easily just extend the default container and add some functionality without needing to re-setup the work that Slim has already done.

What's the benefit of declaring `final` on an open source framework (no sarcasm).

**Comment by akrabat** [comment]
https://ocramius.github.io/blog/when-to-declare-classes-final/

**Comment by silentworks** [comment]
@stevenwadejr I would like to know your use case for extending the Container rather than replacing it?

The benefit of declaring `final` is making sure we don't have crazy inheritance chains and having support issues due to someone breaking their code due to wrongful use of inheritance. With `final` and an `interface` that defines the external API, we have less chances of this happening.

The only against I can see for the `Container` class is that you will need to redefine the services registered, but this is more due to the fact that `set` method is not part of container-interop.

In a way we have gone overzealous on flexibility in this version of Slim, but at the same time we had to keep a balance, we want you to be able to replace the `Container` but not extend it. Pimple is meant to be simple, I think at the point you want to make it more than simple you should probably replace it.

**Comment by stevenwadejr** [comment]
@silentworks, specifically I wanted to add a `make` method, similar to Laravel's, to the container. I was going to play off of your blog post and ADR and have Actions that had dependencies automatically resolved and injected when resolved via the router. The idea was to just extend and add that functionality to the container rather than having to bootstrap the service bindings again. 

I'm sure there are many other uses for extending the container rather than creating a new one. I would just think having `final` on such a minimalistic framework would limit users. Just an opinion though. Obviously there are work arounds, but just wanted to throw my 2 cents in.

**Comment by designermonkey** [comment]
I have always been a disliker of the `final` and `private` keywords, but more recently I have started to realise, as stated above, that having a decent interface and using a composition over inheritance design plan is better than lots of inheritance. To me, `final` is used right in this scenario, and I have also swapped out the `CallableResolver` myself.

I personally, would opt for more `interfaces` and `traits`, allowing code to be split out into reusable chunks that developers could use, but I can't see many use cases for that pattern in Slim as it stands.
